code/edgemedia/src/main/java/com/adobe/marketing/mobile/edge/media/internal/MediaEventProcessor.kt

```diff
@@ -11,18 +11,149 @@
+    }
+
+    @VisibleForTesting
+    internal val mediaSessions: MutableMap<String, MediaSession> = mutableMapOf()
+
+    /**
+     * Creates new [MediaSession], assigning it a new random ID.
+     * @return the session ID of the new [MediaSession]
+     */
+    fun createSession(): String {
+        synchronized(sessionsMutex) {
+            val sessionId = UUID.randomUUID().toString()
+            val session = MediaRealTimeSession(sessionId, mediaState, dispatcher)
+            mediaSessions[sessionId] = session
+            Log.trace(LOG_TAG, sourceTag, "Created new session ($sessionId)")
+            return sessionId
+        }
+    }
+
+    /**
+     * Ends the [MediaSession] with the given `sessionId`.
+     * @param sessionId the ID of the [MediaSession] to end
+     * @see [MediaSession.end]
+     */
+    fun endSession(sessionId: String) {
+        synchronized(sessionsMutex) {
+            val session = mediaSessions[sessionId]
+            if (session != null) {
+                session.end()
+                if (isMediaSessionInactive(session)) {
+                    mediaSessions.remove(sessionId)
+                }
+            } else {
+                Log.trace(
+                    LOG_TAG,
+                    sourceTag,
+                    "Cannot end media session as session ID ($sessionId) is invalid."
+                )
+            }
+        }
+    }
+
+    /**
+     * Queues the [XDMMediaEvent] for processing in the [MediaSession] with ID `sessionId`.
+     * @param sessionId the ID of the [MediaSession] to queue the event
+     * @param event the [XDMMediaEvent] to process
+     * @see [MediaSession.queue]
+     */
+    fun processEvent(sessionId: String, event: XDMMediaEvent) {
+        synchronized(sessionsMutex) {
+            val session = mediaSessions[sessionId]
+            if (session != null) {
+                session.queue(event)
+                Log.trace(
+                    LOG_TAG,
+                    sourceTag,
+                    "Successfully queued event (${event.xdmData.eventType}) for session ($sessionId)"
+                )
+            } else {
+                Log.trace(
+                    LOG_TAG,
+                    sourceTag,
+                    "Cannot queue event (${event.xdmData.eventType}) as session ID ($sessionId) is invalid."
+                )
+            }
+        }
+    }
+
+    /**
+     * Sets the `backendSessionId` to the [MediaSession] with matching `requestEventId`.
+     * @see [MediaSession.handleSessionUpdate]
+     */
     fun notifyBackendSessionId(requestEventId: String, backendSessionId: String?) {
+        synchronized(sessionsMutex) {
+            mediaSessions.forEach { (_, session) ->
+                session.handleSessionUpdate(requestEventId, backendSessionId)
+            }
+
+            // Session may be aborted if backend session ID is invalid
+            mediaSessions.values.removeAll { isMediaSessionInactive(it) }
```

Comment:
`endSession()` might not be able to delete session since it might have few events to be sent. That's why I was suggesting we use a clear method which does the inactive check for all the sessions and removes them, like we do for notifyBackendSessionId and notifyErrorResponse methods. That way also don't need extra function var to check since it will be a condition inside that function. And we also have just one way of clearing sessions that is bulk and which will take care of all the sessions which may have been ended but cleared.